File: visitor-pattern/java/src/ConcreteVisitorA.java
package src;

// 具体访问者A
public class ConcreteVisitorA implements Visitor {

  // 假如由不同厂商是程序的访问者
  private String name = "Google Visitor";

  @Override
  public void visit(ConcreteElementA concreteElementA) {
    System.out.println(
        "ConcreteVisitorA::visit() [Element.class = " + concreteElementA.getClass().getSimpleName()
            + " Element.name = "
            + concreteElementA.getName() + "]");
    concreteElementA.operate();
  }

  @Override
  public void visit(ConcreteElementB concreteElementB) {
    System.out.println("ConcreteVisitorA::visit() [Element.class = " + concreteElementB.getClass().getSimpleName()
        + " Element.name = "
        + concreteElementB.getName() + "]");
    concreteElementB.operate();
  }

  public String getName() {
    return this.name;
  }
}

File: visitor-pattern/ts/src/ConcreteVisitorA.ts
import { Visitor } from './Visitor.js'
import { ConcreteElementA } from './ConcreteElementA.js'
import { ConcreteElementB } from './ConcreteElementB.js'

// 具体访问者A
export class ConcreteVisitorA implements Visitor {
  name: string
  constructor() {
    // 假如由不同厂商是程序的访问者
    this.name = 'Google Visitor'
  }

  // TS实现函数重载
  visit(concreteElement: ConcreteElementA): void
  visit(concreteElement: ConcreteElementB): void
  visit(concreteElement: ConcreteElementA | ConcreteElementB) {
    console.log(
      'ConcreteVisitorA::visit() [Element.class = ' +
        concreteElement.constructor.name +
        ' Element.name = ' +
        concreteElement.getName() +
        ']'
    )
    concreteElement.operate()
  }

  getName() {
    return this.name
  }
}
